amd_openvx_extensions/amd_media/decoder.cpp

```diff
@@ -762,14 +840,55 @@ void CLoomIoMediaDecoder::DecodeLoop(int mediaIndex)
                     if(err) {
                         fprintf(stderr,"map output for sw_scale: clFinish failed (%d)",  err);
                     }
+#elif ENABLE_HIP
+                    uint8_t *dst_data[4] = {0};
+                    int dst_linesize[4] = {0};
+                    dst_data[0] = decodeBuffer[bufId];
+                    dst_linesize[0] = gpuStride;
+                    if (outputFormat == AV_PIX_FMT_NV12) {
+                        dst_data[1] = decodeBuffer[bufId] + decoderImageHeight * gpuStride;
+                        dst_linesize[1]  = gpuStride;
+                    }
+                    // do sws_scale
+                    int ret = sws_scale(conversionContext[mediaIndex], tmp_frame->data, tmp_frame->linesize, 0, tmp_frame->height, dst_data, dst_linesize);
+                    if (ret < decoderImageHeight) {
+                        fprintf(stderr, "Error in output image scaling using sws_scale\n");
+                        continue;
+                    }
+                    #if DUMP_DECODED_FRAME
+                    if (fpIn) {
+                        fwrite(dst_data[0], 1, decoderImageHeight * dst_linesize[0], fpIn);
+                        if (outputFormat == AV_PIX_FMT_NV12)
+                            fwrite(dst_data[1], 1, (decoderImageHeight >> 1) * dst_linesize[1], fpIn);
+                    }
+                    #endif
+                    if (!hip_dev_prop.canMapHostMemory) {
+                        hipError_t err = hipMemcpyHtoD((void *)((uint8_t *)mem[bufId] + gpuOffset + mediaIndex * mapHeight * gpuStride), decodeBuffer[bufId], mapHeight * gpuStride);
+                        if (err != hipSuccess) {
+                            vxAddLogEntry((vx_reference)node, VX_FAILURE, "ERROR: hipMemcpyHtoD(buf[%d], slice[%d]) failed (%d)\n", bufId, mediaIndex, err);
+                            continue;
+                        }
+                    }
+#endif
+
                 } else {
                     // copy AV frame to output
-                    cl_int err = clEnqueueWriteBuffer(cmdq, mem[bufId], CL_TRUE, clOffset + mediaIndex * decoderImageHeight * clStride, decoderImageHeight * clStride, tmp_frame->data, 0, nullptr, nullptr);
+#if ENABLE_OPENCL
+                    cl_int err = clEnqueueWriteBuffer(cmdq, (cl_mem)mem[bufId], CL_TRUE, gpuOffset + mediaIndex * decoderImageHeight * gpuStride, decoderImageHeight * gpuStride, tmp_frame->data, 0, nullptr, nullptr);
                     if (err < 0) {
                         vxAddLogEntry((vx_reference)node, VX_FAILURE, "ERROR: clEnqueueWriteBuffer(buf[%d], slice[%d]) failed (%d)\n", bufId, mediaIndex, err);
                         continue;
                     }
                     clFinish(cmdq);
+#elif ENABLE_HIP
+                    if (!hip_dev_prop.canMapHostMemory) {
```

Comment:
Looks like Line#866 is already copying the data to device. Do we need to do it again here?

Reply:
yes, these are for different conditions. line#866 is for this condition if (conversionContext[mediaIndex] != NULL) {
and line#885 is for the else condition..